transformer_engine/paddle/layer/attention.py

```diff
@@ -458,6 +464,24 @@ def __init__(
 
         self.use_fused_attention = bool(int(os.getenv("NVTE_FUSED_ATTN", "1")))
 
+        self.deterministic = not bool(int(os.getenv("NVTE_ALLOW_NONDETERMINISTIC_ALGO", "1")))
+
+        # CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT
+        # - unset:       enables workspace optimization when required workspace is <= 256MB
+        #                or when bias gradient needs to be computed
+        # - n:           enables workspace optimization when required workspace is <= n bytes
+        # - -1:          enables workspace optimization always
+        # - 0:           disables workspace optimization always
+        if "NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT" in os.environ:
+            if os.environ["NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT"] == "0":
+                os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "0"
+            if os.environ["NVTE_FUSED_ATTN_FORCE_WORKSPACE_OPT"] == "1":
+                os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "-1"
+
+        if self.deterministic:
+            # workspace optimization path is deterministic
+            os.environ["CUDNN_FRONTEND_ATTN_DP_WORKSPACE_LIMIT"] = "-1"
+
```

Comment:
Maybe add a note here about the cuDNN versions? Something like this? https://github.com/NVIDIA/TransformerEngine/blob/8e039fdcd98fc56582d81e373880c1509c2b8f73/transformer_engine/pytorch/attention.py#L5148C1-L5150C72

Also, could you please pass the `self.deterministic` value down to `nvte_fused_attn_bwd_xxx` calls?  Thanks.

Reply:
Done. BTW, I changed the `if (cudnn_runtime_version >= 90000 && sm_arch_ >= 90)` to `if (cudnn_runtime_version >= 90000)` here[fused_attn_f16_arbitrary_seqlen.cu](https://github.com/NVIDIA/TransformerEngine/pull/956/files#diff-28710e6e65e3526663810821d4131eaeccb3e3807262a8fbd9ecfa5799739cfb)
. `set_deterministic_algorithm` is also applicable and essential for Ada and Ampere.